### PR TITLE
feat(ui): add the ability to clone an environment

### PR DESCRIPTION
**Pull request by bnjjj**
tested, to merge
Signed-off-by: Benjamin Coenen <benjamin.coenen@hotmail.com>

### CODE REVIEW

File: ui/src/app/views/project/show/environment/list/item/environment.component.ts [hunk 1]
```diff
@@ -42,6 +44,21 @@ export class ProjectEnvironmentComponent {
...
+            return;
+        }
+
+        this.cloneLoading = true;
+
+        this._projectStore.cloneProjectEnvironment(this.project.key, this.editableEnvironment, this.cloneName)
+            .finally(() => this.cloneLoading = false)
+            .subscribe(() => {
+                this._toast.success('', this._translate.instant('environment_cloned'));
+            });
```
Comment (sguiheux) [inline ui/src/app/views/project/show/environment/list/item/environment.component.ts]: CloneModal.hide?

File: ui/src/assets/i18n/en.json [hunk 2]
```diff
@@ -150,11 +150,15 @@
...
   "common_yes" : "Yes",
+  "common_clone" : "Clone",
 
   "danger_zone" : "Danger Zone",
 
   "environment_created" : "Environment added",
   "environment_create_placeholder" : "New environment",
+  "environment_cloned" : "Environment cloned",
+  "environment_title_clone" : "Clone the environment",
+  "environment_clone_placeholder" : "Name of the new environnement",
```
Comment (sguiheux) [inline ui/src/assets/i18n/en.json]: French word :-)

File: engine/api/environment.go [hunk 3]
```diff
@@ -429,31 +429,19 @@ func updateEnvironmentHandler(w http.ResponseWriter, r *http.Request, db *gorp.D
...
 	vars := mux.Vars(r)
 	projectKey := vars["key"]
 	environmentName := vars["permEnvironmentName"]
+	cloneName := vars["cloneName"]
 
 	env, errEnv := environment.LoadEnvironmentByName(db, projectKey, environmentName)
 	if errEnv != nil {
-		log.Warning("cloneEnvironmentHandler> Cannot load environment %s: %s\n", environmentName, errEnv)
-		return errEnv
+		return sdk.WrapError(errEnv, "cloneEnvironmentHandler> Cannot load environment %s: %s\n", environmentName, errEnv)
```
Comment (yesnault) [inline engine/api/environment.go]: remove \n

File: engine/api/environment.go [hunk 4]
```diff
@@ -429,31 +429,19 @@ func updateEnvironmentHandler(w http.ResponseWriter, r *http.Request, db *gorp.D
...
-
-	var envPost sdk.Environment
-	if err := UnmarshalBody(r, &envPost); err != nil {
-		return err
-	}
-
-	//Check if the new environment has a name
-	if envPost.Name == "" {
-		return sdk.ErrWrongRequest
+		return sdk.WrapError(errProj, "cloneEnvironmentHandler> Cannot load project %s: %s\n", projectKey, errProj)
```
Comment (yesnault) [inline engine/api/environment.go]: remove \n

File: engine/api/environment.go [hunk 5]
```diff
@@ -463,64 +451,61 @@ func cloneEnvironmentHandler(w http.ResponseWriter, r *http.Request, db *gorp.Db
...
-			return err
+			return sdk.WrapError(err, "cloneEnvironmentHandler> Unable to insert group in environment: %s", err)
 		}
 	}
 
 	//Update the poroject
 	if err := project.UpdateLastModified(tx, c.User, p); err != nil {
-		log.Warning("cloneEnvironmentHandler> Cannot update last modified date: %s\n", err)
-		return err
+		return sdk.WrapError(err, "cloneEnvironmentHandler> Cannot update last modified date: %s\n", err)
```
Comment (yesnault) [inline engine/api/environment.go]: remove \n

File: engine/api/environment.go [hunk 6]
```diff
@@ -463,64 +451,61 @@ func cloneEnvironmentHandler(w http.ResponseWriter, r *http.Request, db *gorp.Db
...
 	}
 
-	//return the project wil all environment
+	//return the project with all environments
 	var errEnvs error
 	p.Environments, errEnvs = environment.LoadEnvironments(db, p.Key, true, c.User)
 	if errEnvs != nil {
-		log.Warning("cloneEnvironmentHandler> Cannot load environments: %s\n", errEnvs)
-		return errEnvs
+		return sdk.WrapError(errEnvs, "cloneEnvironmentHandler> Cannot load environments: %s\n", errEnvs)
```
Comment (yesnault) [inline engine/api/environment.go]: remove \n

File: sdk/environment.go [hunk 7]
```diff
@@ -148,14 +148,8 @@ func DeleteEnvironment(pk, name string) error {
...
 func CloneEnvironment(pk, name, new string) (*Environment, error) {
-	var e = Environment{Name: new}
-	b, err := json.Marshal(e)
-	if err != nil {
-		return nil, err
-	}
-
-	path := fmt.Sprintf("/project/%s/environment/%s/clone", pk, name)
-	if _, _, err := Request("POST", path, b); err != nil {
+	path := fmt.Sprintf("/project/%s/environment/%s/clone/%s", pk, name, new)
```
Comment (yesnault) [inline sdk/environment.go]: queryEscape(..) on new

**Comment by fsamin** [comment]
👍 
I think there is a bug on API side. It doesn't clone password environment with the right value.
Well i don't know if it's a bug or a feature :/

**Comment by yesnault** [comment]
for Type Password -> not cloned. It's a feature, do avoid human error after cloning a production environment on another environment...

**Pull request merged**